locked-nft/contracts/LockedNFT.cdc

```diff
@@ -0,0 +1,131 @@
+import NonFungibleToken from "./NonFungibleToken.cdc"
+
+pub contract LockedNFT {
+
+    pub event ContractInitialized()
+    pub event NFTLocked(
+        id: UInt64,
+        to: Address?,
+        lockedAt: UInt64,
+        lockedUntil: UInt64
+    )
+    pub event NFTUnlocked(id: UInt64, from: Address?)
+
+    pub let CollectionStoragePath:  StoragePath
+    pub let CollectionPublicPath:   PublicPath
+
+    pub var totalLockedTokens:      UInt64
+    access(self) let lockedTokens:  {UInt64: LockedData}
+
+    pub struct LockedData {
+        pub let owner: Address
+        pub let lockedAt: UInt64
+        pub let lockedUntil: UInt64
+        pub let nftType: String
+
+        init (id: UInt64, owner: Address, duration: UInt64, nftType: String) {
+            if let lockedToken = LockedNFT.lockedTokens[id] {
+                self.owner = lockedToken.owner
+                self.lockedAt = lockedToken.lockedAt
+                self.lockedUntil = lockedToken.lockedUntil
+                self.nftType = lockedToken.nftType
+            } else {
+                self.owner = owner
+                self.lockedAt = UInt64(getCurrentBlock().timestamp)
+                self.lockedUntil = self.lockedAt + duration
+                self.nftType = nftType
+            }
+        }
+    }
+
+    pub fun getLockedToken(id: UInt64): LockedNFT.LockedData? {
+        return LockedNFT.lockedTokens[id]
+    }
+
+    pub fun canUnlockToken(id: UInt64): Bool {
+        if let lockedToken = LockedNFT.lockedTokens[id] {
+            if lockedToken.lockedUntil < UInt64(getCurrentBlock().timestamp) {
+                return true
+            }
+        }
+
+        return false
+    }
+
+    pub resource interface LockedCollection {
+        pub fun getIDs(): [UInt64]
+        pub fun borrowNFT(id: UInt64): &NonFungibleToken.NFT
+        pub fun lock(token: @NonFungibleToken.NFT, duration: UInt64)
+        pub fun unlock(id: UInt64): @NonFungibleToken.NFT
+    }
+
+    pub resource Collection: LockedCollection {
+        pub var lockedNFTs: @{UInt64: NonFungibleToken.NFT}
+
+        pub fun unlock(id: UInt64): @NonFungibleToken.NFT {
+            pre {
+                LockedNFT.canUnlockToken(
+                    id: id
+                ) == true : "locked duration has not been met"
+            }
+
+            let token <- self.lockedNFTs.remove(key: id) ?? panic("Missing NFT")
+            LockedNFT.lockedTokens.remove(key: id)
+            LockedNFT.totalLockedTokens = LockedNFT.totalLockedTokens - 1
+            emit NFTUnlocked(id: token.id, from: self.owner?.address)
+
+            return <-token
+        }
+
+        pub fun lock(token: @NonFungibleToken.NFT, duration: UInt64) {
+            let id: UInt64 = token.id
+            let oldToken <- self.lockedNFTs[id] <- token
```

Comment:
could the same collection end up being used for different nft types? if yes, this will cause a collision

Reply:
Thank you Deewai. I will resolve.